From aa00707d3d9ba535c932bb4d9f0edab1ec79f94d Mon Sep 17 00:00:00 2001
From: Sonali Naresh <sonanare@qti.qualcomm.com>
Date: Thu, 18 Jun 2026 11:37:54 +0530
Subject: [PATCH] WSC M1 has WSC Model Name TLV with incorrect length Fix is to
 resize it appropriately. Issue was seen earlier in MAUT tests. Now its seen
 with MCUT

Signed-off-by: Sonali Naresh <sonanare@qti.qualcomm.com>
---
 .../src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp | 4 ++--
 controller/src/beerocks/master/controller.cpp              | 7 +++++++
 framework/tlvf/src/include/tlvf/WSC/m1.h                   | 7 +++++++
 3 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 46fc99d..3340e24 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -1496,8 +1496,8 @@ bool ApAutoConfigurationTask::add_wsc_m1_tlv(const std::string &radio_iface)
     cfg.device_name         = "prplmesh-agent";
 
     //WSC M1 packet Model name has size restrictions
-    if (cfg.model_name.size() > 30) {
-        cfg.model_name.resize(30);
+    if (cfg.model_name.size() > WSC_MODEL_NAME_MAX_LENGTH) {
+        cfg.model_name.resize(WSC_MODEL_NAME_MAX_LENGTH);
     }
 
     switch (radio->wifi_channel.get_freq_type()) {
diff --git a/controller/src/beerocks/master/controller.cpp b/controller/src/beerocks/master/controller.cpp
index eb57020..84c888a 100755
--- a/controller/src/beerocks/master/controller.cpp
+++ b/controller/src/beerocks/master/controller.cpp
@@ -1113,6 +1113,13 @@ bool Controller::autoconfig_wsc_add_m2(WSC::m1 &m1,
     beerocks::bpl::get_model_name(m2_cfg.model_name);
     beerocks::bpl::get_model_number(m2_cfg.model_number);
     beerocks::bpl::get_serial_number(m2_cfg.serial_number);
+
+    // WSC M2 packet Model Name has size restrictions. A too-long Model Name produces a
+    // malformed M2, so truncate it to the maximum allowed length.
+    if (m2_cfg.model_name.size() > WSC_MODEL_NAME_MAX_LENGTH) {
+        m2_cfg.model_name.resize(WSC_MODEL_NAME_MAX_LENGTH);
+    }
+
     m2_cfg.primary_dev_type_id = WSC::WSC_DEV_NETWORK_INFRA_GATEWAY;
     m2_cfg.device_name         = "prplmesh-controller";
     m2_cfg.encr_type_flags     = uint16_t(WSC::eWscEncr::WSC_ENCR_NONE) |
diff --git a/framework/tlvf/src/include/tlvf/WSC/m1.h b/framework/tlvf/src/include/tlvf/WSC/m1.h
index a4cee50..b8fb89d 100644
--- a/framework/tlvf/src/include/tlvf/WSC/m1.h
+++ b/framework/tlvf/src/include/tlvf/WSC/m1.h
@@ -11,6 +11,13 @@
 
 #include <tlvf/WSC/WscAttrList.h>
 
+/**
+ * Maximum allowed length (in bytes) of the WSC Model Name attribute.
+ * A Model Name longer than this produces a malformed WSC M1/M2 message, so it must be
+ * truncated to this length before being placed into the WSC TLV.
+ */
+#define WSC_MODEL_NAME_MAX_LENGTH 30
+
 namespace WSC {
 
 class m1 : public WscAttrList {
-- 
2.34.1

